### PR TITLE
Adding Uptime tracking capability

### PR DESCRIPTION
**Pull request by chajain**
## Description

Uptime data is included in the data sent to SCC via RMT
https://jira.suse.com/browse/CSE-504

Fixes # (issue)

Fixes # (issue)

## Change Type

*Please select the correct option.*

- [ ] **Bug Fix** (a non-breaking change which fixes an issue)
- [X] **New Feature** (a non-breaking change which adds new functionality)
- [ ] **Documentation Update** (a change which only updates documentation)

## Checklist

*Please check off each item if the requirement is met.*

- [X] I have verified that my code follows RMT's coding standards with `rubocop`.
- [ ] I have reviewed my own code and believe that it's ready for an external review.
- [ ] I have provided comments for any hard-to-understand code.
- [ ] I have documented the `MANUAL.md` file with any changes to the user experience.
- [X] RMT's test coverage remains at 100%.
- [ ] If my changes are non-trivial, I have added a changelog entry to notify users at `package/obs/rmt-server.changes`.

## Other Notes

Please use this space to provide notes or thoughts to the team, such as tips on how to review/demo your changes.
How to test this PR

On the client system clone https://github.com/SUSE/connect-ng/pull/202 ( this PR has the connect-ng changes for uptime tracking feature )
Also clone https://github.com/SUSE/uptime-tracker
On the RMT system, clone this code
Execute rails db:migrate to create the system_uptimes table
Start the server bin/rails server -b 0.0.0.0
Register the system on client system using command suseconnect --url <rmt_system_ip> , it should populate the system_uptimes table with the entries from /etc/zypp/suse-uptime.logs on client system

### CODE REVIEW

**Review by ngetahun — CHANGES_REQUESTED** [review]

File: db/migrate/20240111200053_create_system_uptimes.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+class CreateSystemUptimes < ActiveRecord::Migration[6.1]
+  def change
+    safety_assured do
+      create_table :system_uptimes, id: :serial, force: :cascade do |t|
```
[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/migrate/20240111200053_create_system_uptimes.rb:4]: As per the doc https://apidock.com/rails/v5.2.3/ActiveRecord/ConnectionAdapters/SchemaStatements/create_table
:force
Set to true to drop the table before creating it. Set to :cascade to drop dependent objects as well. Defaults to false.
:id
Whether to automatically add a primary key column. Defaults to true. Join tables for {ActiveRecord::Base.has_and_belongs_to_many}[rdoc-ref:Associations::ClassMethods#has_and_belongs_to_many] should set it to false.

A Symbol can be used to specify the type of the generated primary key column.

[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/migrate/20240111200053_create_system_uptimes.rb:4]: I tried to remove as suggested
`class CreateSystemUptimes < ActiveRecord::Migration[6.1]
  def change
    safety_assured do
      create_table :system_uptimes do |t|
        t.bigint :system_id, null: false
        t.date :online_at_day, null: false
        t.column :online_at_hours, 'binary(24)', null: false
        t.timestamps
      end

      commit_db_transaction

      add_index :system_uptimes, %i[system_id online_at_day], unique: true, name: 'id_online_day'

      add_foreign_key :system_uptimes, :systems, column: :system_id, validate: false
    end
  end
end` db migrate failed

[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/migrate/20240111200053_create_system_uptimes.rb:4]:
```
StandardError: An error has occurred, all later migrations canceled:

Table 'system_uptimes' has no foreign key for systems
/home/vagrant/rmt/bin/rails:9:in `<top (required)>'
/home/vagrant/rmt/bin/spring:15:in `<top (required)>'
```

[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/migrate/20240111200053_create_system_uptimes.rb:4]: Not sure we need to remove id. Please clarify.

File: package/files/systemd/rmt-uptime-cleanup.timer [hunk 2]
```diff
@@ -0,0 +1,10 @@
+[Unit]
+Description=RMT Uptime Data cleanup timer
+
+[Timer]
+# Run this timer every day at a randomized 24h delay.
+OnCalendar=daily
```
[the remark this answers is not in the document]
Comment (yarunachalam) [reply package/files/systemd/rmt-uptime-cleanup.timer]: done

File: db/migrate/20240111200053_create_system_uptimes.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/migrate/20240111200053_create_system_uptimes.rb]: done

File: db/schema.rb [hunk 3]
```diff
@@ -12,172 +12,182 @@
 
 ActiveRecord::Schema.define(version: 2024_01_29_140413) do
```
[the remark this answers is not in the document]
Comment (yarunachalam) [reply db/schema.rb]: done

**Comment by rjschwei** [comment]
AFIK version bumps are handled independently, i.e.  lib/rmt.rb and the spec file changes should not be part of the PR. The Public Cloud team expects support for Debian repositories as the primary feature of 2.16, not sure is that is merged yet.

**Comment by digitaltom** [comment]
> AFIK version bumps are handled independently, i.e. lib/rmt.rb and the spec file changes should not be part of the PR. 
> The Public Cloud team expects support for Debian repositories as the primary feature of 2.16, not sure is that is merged yet. 
Yes, please just add a note about the changes to the changelog, we'll include it into the next release and take care about the versioning.

**Comment by yarunachalam** [comment]
> > AFIK version bumps are handled independently, i.e. lib/rmt.rb and the spec file changes should not be part of the PR.
> > The Public Cloud team expects support for Debian repositories as the primary feature of 2.16, not sure is that is merged yet.
> 
> Yes, please just add a note about the changes to the changelog, we'll include it into the next release and take care about the versioning.

Do I need to remove the spec file from this PR ?

**Comment by digitaltom** [comment]
> Do I need to remove the spec file from this PR ?

I think you have changes in the .spec file that you need to keep regarding the timers. But you don't need to change the version number. Just add a description of the change to the [changelog](https://github.com/SUSE/rmt/blob/master/package/obs/rmt-server.changes), and the team will take care about the versioning + release.

File: package/obs/rmt-server.changes [hunk 4]
```diff
@@ -1,3 +1,8 @@
+- Adding Uptime tracking capability
+  * Uptime data is included in the data sent to SCC via RMT
```
Comment (ngetahun) [inline package/obs/rmt-server.changes:2]: Please add here the relevant bugzilla(bsc)/jira(ped) tickets here. We can cleanup the rest when preparing for rmt 2.17.

File: lib/suse/connect/system_serializer.rb [hunk 5]
```diff
@@ -47,6 +48,17 @@ def products
     end
   end
 
+  def online_at
+    object.system_uptimes.map do |systemuptime|
+      payload = {
+        system_id: systemuptime.system_id,
```
Comment (digitaltom) [inline lib/suse/connect/system_serializer.rb:54]: Afaics that parameter is not used in https://github.com/SUSE/happy-customer/pull/7339/, what's the reason for including it into the payload?

Comment (yarunachalam) [reply lib/suse/connect/system_serializer.rb:54]: fixed

**Review by Adnilson — CHANGES_REQUESTED** [review]
We can correct some variable inconsistency.

There is `system_uptime` and sometimes `systemuptime`.

Can we have this changed? It also helps with reading.

**Comment by chajain** [comment]
Performed the end to end testing along with SCC latest PR https://github.com/SUSE/happy-customer/pull/7339 and connect-ng PR https://github.com/SUSE/connect-ng/pull/210 and verified following scenarios

- Creation of system_uptime table with db:migrate 
- Populating the system_uptime table when the keepalive heartbeat is sent from client
- `rmt-cli systems scc-sync` syncs the system_uptime tables on SCC from RMT

File: spec/lib/suse/connect/system_serializer_spec.rb [hunk 6]
```diff
@@ -86,4 +86,13 @@
       expect(serializer.key? :system_token).to eq(false)
     end
   end
+
+  context 'system with systemuptime' do
```
Comment (Adnilson) [inline spec/lib/suse/connect/system_serializer_spec.rb:90]: I would also change the description but I don't see it as a blocker.

**Review by Adnilson — APPROVED** [review]

**Pull request merged**